### PR TITLE
Fix for issue #4708

### PR DESCRIPTION
**Pull request by PedroTadim**
This is my fix for issue #4708. Now running tests. The UTF-8 analyze function became more complex, can this create a performance degradation for scanning csv and parquet files?

### CODE REVIEW

**Comment by Mytherin** [comment]
Thanks for the PR! Looks great. This only adds a few bitwise operations so I don't expect any major regressions - moving strings around will likely still be the dominating factor here. Could we perhaps add a few more tests and ensure we get full branch coverage on the function so that each invalid unicode path is triggered in at least one test?

**Review by Mytherin — APPROVED** [review]
Thanks! LGTM

**Comment by PedroTadim** [comment]
There's performance degradation. I was thinking to unroll the branches for the 2,3 and 4 byte sequence cases.

**Comment by Mytherin** [comment]
Feel free to have a go at optimizing it. It is certainly a piece of code that is run many times. I would say optimizing for the mostly-ASCII case is the most worthwhile pursuit here.

File: third_party/utf8proc/utf8proc_wrapper.cpp [hunk 1]
```diff
@@ -30,71 +30,71 @@ static void AssignInvalidUTF8Reason(UnicodeInvalidReason *invalid_reason, size_t
 	}
 }
 
+#define EXTRA_BYTE_LOOP(NEXTRA_BYTES) \
```
Comment (Mytherin) [inline third_party/utf8proc/utf8proc_wrapper.cpp:33]: Can we avoid defines and use templates instead? :)

Comment (PedroTadim) [reply third_party/utf8proc/utf8proc_wrapper.cpp:33]: I was predicting these :)

**Comment by Tishj** [comment]
> There's performance degradation. I was thinking to unroll the branches for the 2, 3 and 4 byte sequence cases.

Shouldn't it be able to unroll them if you make the `nextra_bytes` a template parameter, so it's known at compile time?

**Comment by Mytherin** [comment]
As this is such an important function, I decided to write a script to verify that the behavior matches that of Python by essentially generating all pairs of byte sequences of length [1, 2] and sampling byte sequences of length [3, 4, 5]. Then running both Python's decode and DuckDB's decode and verifying that they both either succeed or both fail.  

```pyimport duckdb
import random
con = duckdb.connect()
con.execute('PRAGMA disable_optimizer')
from tqdm import tqdm

# sample rate: higher means more values are sampled
# this represents approximately how many values we sample from each domain
sample_rate = 100000

f = open('results.csv', 'w+')

for byte_count in range(1, 6):
	total_byte_count = 256 ** byte_count
	entry_skip = total_byte_count // sample_rate
	if entry_skip > 0:
		total_byte_count = total_byte_count // entry_skip
	for entry in tqdm(range(total_byte_count)):
		# decompose the entry into byte_count different bytes
		if entry_skip > 0:
			entry *= entry_skip
			entry += int(random.random() * entry_skip)
		bytes = []
		for byte_index in range(byte_count):
			null_byte = False
			byte = (entry >> (8 * byte_index)) % 256
			if byte == 0:
				# skip null bytes
				null_byte = True
				break
			bytes.append(byte)
		if null_byte:
			continue

		# decode the byte in python
		byte_string = bytearray(bytes)
		decode_success = True
		try:
			byte_string.decode('utf8')
		except:
			decode_success = False

		# now decode the byte in duckdb
		hex_bytes = ''.join([('\\x0' + hex(byte)[2:] if byte < 16 else '\\x' + hex(byte)[2:]) for byte in bytes])

		duckdb_decode_success = True
		try:
			con.execute(f"SELECT decode('{hex_bytes}'::BLOB) IS NOT NULL").fetchall()
		except:
			duckdb_decode_success = False
		if decode_success != duckdb_decode_success:
			f.write(f"Error in byte string {hex_bytes}\n")
			f.write(f"Python decode result: {'Success' if decode_success else 'Fail'}\n")
			f.write(f"DuckDB decode result: {'Success' if duckdb_decode_success else 'Fail'}\n")
			f.write('\n')

f.close()

```

There are still a number of inconsistencies between the Python decode and the DuckDB decode, for example:

```
Error in byte string \xc0\x80
Python decode result: Fail
DuckDB decode result: Success

Error in byte string \xc1\x80
Python decode result: Fail
DuckDB decode result: Success

Error in byte string \xc0\x81
Python decode result: Fail
DuckDB decode result: Success

Error in byte string \xc1\x81
Python decode result: Fail
DuckDB decode result: Success

Error in byte string \xc0\x82
Python decode result: Fail
DuckDB decode result: Success

Error in byte string \xc1\x82
Python decode result: Fail
DuckDB decode result: Success
...
```

Running this in Python:

```
>>> b'\xc0\x80'.decode('utf8')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
UnicodeDecodeError: 'utf-8' codec can't decode byte 0xc0 in position 0: invalid start byte

```

Whereas in DuckDB this still appears to succeed:

```
D select decode('\xc0\x80'::blob);
┌──────────────────────────┐
│ decode('\xC0\x80'::BLOB) │
├──────────────────────────┤
│ ??                         │
└──────────────────────────┘
```

Could you perhaps have a look at those?

**Comment by PedroTadim** [comment]
Sure I will do it

**Comment by Mytherin** [comment]
Thanks for the fixes! This seems to pass my test now. Good stuff!
